app/models/manageiq/providers/ansible_tower/shared/inventory/parser/automation_manager.rb

```diff
@@ -74,27 +74,33 @@ def configuration_script_sources
     end
   end
 
+  def tower_credential_type(tower_type)
+    @provider_module ||= ManageIQ::Providers::Inflector.provider_module(collector.manager.class).name
+    @tower_credential_type ||= begin
```

Comment:
D'oh, you changed the wrong thing ... I guess I wasn't clear.

I was thinking that `miq_credential_type` was fine for the method name.  But, the *param* name should change to `tower_credential_type` for symmetry.

However, read on...

I was confused by this ivar name matching the method name because the ivar is an array, but the method returns a single value.

Here's what I suggest doing.

1. Create a separate ivar-bound method called `miq_credential_types` that moves all the code form lines 78-94 in this method (in that method, move `provider_module` inside the `begin` block then you don't have to have another ivar.
2. Then, this method becomes:

```ruby
def miq_credential_type(tower_credential_type)
  miq_credential_types[tower_credential_type] || "#{@provider_module}::AutomationManager::Credential"
end
```

That way, you have one method that just loads the array `miq_credentials_types`, which is bound to an ivar called `@miq_credential_types`.  Then, you have another method that uses that cached list of credential types to look up the appropriate credential, or returns the default.

Reply:
it actually began like that. and then i don't want to keep looking up this `provider_module` again. see my next version...